File: backend/controllers/auth_controller.js
const { getDb } = require('../config/db');
const bcrypt = require('bcrypt');

const signup = async (req, res) => {

    try {

        const { name, phone_no, email, password } = req.body;

        if (!name || !phone_no || !email || !password) {
            return res.status(400).json({ message: "All fields are required" })
        }

        if (password.length < 6) {
            return res.status(400).json({ message: "Password must be at least 6 characters long" });
        }

        const normalizedEmail = email.toLowerCase();

        const db = getDb();

        const existingUser = await db.collection('users').findOne({ email: normalizedEmail });

        if (existingUser) {
            return res.status(400).json({ message: "User already exists" })
        }

        const hasehpassword = await bcrypt.hash(password, 10);

        const newUser = {
            name,
            phone_no,
            email: normalizedEmail,
            password: hasehpassword,
            createdAt: new Date()
        };

        await db.collection('users').insertOne(newUser);

        return res.status(201).json({ message: "User registered successfully" })

    } catch (error) {
        console.error(error);
        return res.status(500).json({ message: "Internal server error" })
    }
}

const login = async (req, res) => {
    try {
        const { email, password } = req.body;

        if (!email || !password) {
            return res.status(400).json({ message: "Email and password are required" });
        }

        const db = getDb();
        const normalizedEmail = email.toLowerCase();

        const user = await db.collection('users').findOne({ email: normalizedEmail });

        if (!user) {
            return res.status(404).json({ message: "User not found. Please signup to continue." });
        }

        const isMatch = await bcrypt.compare(password, user.password);

        if (!isMatch) {
            return res.status(401).json({ message: "Invalid email or password" });
        }

        req.session.user = {
            id: user._id,
            name: user.name,
            email: user.email,
            phone_no: user.phone_no
        };

        return res.status(200).json({
            message: "Login successful",
            user: req.session.user
        });

    } catch (error) {
        console.error(error);
        return res.status(500).json({ message: "Internal server error" });
    }
};

module.exports = { signup, login }

File: backend/python_controllers/auth_controller.py
from fastapi import APIRouter, Request, HTTPException
from passlib.context import CryptContext
from datetime import datetime
from config.db import get_db

router = APIRouter()
pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")


@router.post("/signup")
async def signup(data: dict):
    print("Incoming data:", data)
    name = data.get("name", "").strip()
    phone_no = data.get("phone_no", "").strip()
    email = data.get("email", "").strip()
    password = data.get("password", "")
    confirmPassword = data.get("confirmPassword", "")

    if not all([name, phone_no, email, password, confirmPassword]):
        raise HTTPException(status_code=400, detail="All fields are required")

    if len(password) < 6:
        raise HTTPException(status_code=400, detail="Password must be at least 6 characters")

    if password != confirmPassword :
        raise HTTPException(status_code=400, detail="Password do not match")


    email = email.lower()
    db = get_db()

    if await db.users.find_one({"email": email}):
        raise HTTPException(status_code=400, detail="User already exists")

    hashed_password = pwd_context.hash(password)

    await db.users.insert_one({
        "name": name,
        "phone_no": phone_no,
        "email": email,
        "password": hashed_password,
        "createdAt": datetime.utcnow()
    })

    return {"message": "User registered successfully"}


@router.post("/login")
async def login(request: Request, data: dict):
    email = data.get("email", "").strip()
    password = data.get("password", "")

    if not email or not password:
        raise HTTPException(status_code=400, detail="Email and password are required")

    email = email.lower()
    db = get_db()

    user = await db.users.find_one({"email": email})

    # FIX: Use the same error message for both "not found" and "wrong password"
    # to prevent user enumeration attacks.
    if not user or not pwd_context.verify(password, user["password"]):
        raise HTTPException(status_code=401, detail="Invalid email or password")

    request.session["user"] = {
        "id": str(user["_id"]),
        "name": user["name"],
        "email": user["email"],
        "phone_no": user["phone_no"]
    }

    return {"message": "Login successful", "user": request.session["user"]}


@router.post("/logout")
async def logout(request: Request):
    request.session.clear()
    return {"message": "Logged out"}


@router.get("/me")
async def me(request: Request):
    if "user" not in request.session:
        return {"authenticated": False}

    return {"authenticated": True, "user": request.session["user"]}